presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java

```diff
@@ -0,0 +1,240 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+
+import static java.util.Objects.requireNonNull;
+
+public class ArrayBlockBuilder
+        extends AbstractArrayBlock
+        implements BlockBuilder
+{
+    private final BlockBuilderStatus blockBuilderStatus;
+    private final BlockBuilder values;
+    private final SliceOutput offsets;
+    private final SliceOutput valueIsNull;
+    private int currentEntrySize;
+    private boolean ongoingBlockEntry;
+
+    /**
+     * Caller of this constructor is responsible for making sure `valuesBlock` is constructed with the same `blockBuilderStatus` as the one in the argument
+     */
+    public ArrayBlockBuilder(BlockBuilder valuesBlock, BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this(
+                blockBuilderStatus,
+                valuesBlock,
+                new DynamicSliceOutput((expectedEntries + 1) * 4),
+                new DynamicSliceOutput(expectedEntries));
+    }
+
+    public ArrayBlockBuilder(Type elementType, BlockBuilderStatus blockBuilderStatus, int expectedEntries, int expectedBytesPerEntry)
+    {
+        this(
+                blockBuilderStatus,
+                elementType.createBlockBuilder(blockBuilderStatus, expectedEntries, expectedBytesPerEntry),
+                new DynamicSliceOutput((expectedEntries + 1) * 4),
+                new DynamicSliceOutput(expectedEntries));
+    }
+
+    public ArrayBlockBuilder(Type elementType, BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this(
+                blockBuilderStatus,
+                elementType.createBlockBuilder(blockBuilderStatus, expectedEntries),
+                new DynamicSliceOutput((expectedEntries + 1) * 4),
+                new DynamicSliceOutput(expectedEntries));
+    }
+
+    /**
+     * Caller of this private constructor is responsible for making sure `values` and `offsets` are constructed with the same `blockBuilderStatus` as the one in the argument
+     */
+    private ArrayBlockBuilder(BlockBuilderStatus blockBuilderStatus, BlockBuilder values, SliceOutput offsets, SliceOutput valueIsNull)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+        this.values = requireNonNull(values);
```

Comment:
add second argument for the rest of these, just like the check for `blockBuilderStatus`